src/main/java/jenkins/branch/NameEncoder.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.branch;
+
+/**
+ * Encodes names that are not nice so that they are safe to use as url path segments.
+ * We don't want to do a full url encoding, only replace problematic names with {@code %} escaped variants so
+ * that when double encoded by Stapler etc we bypass issues.
+ *
+ * @since 2.0.0
+ */
+public final class NameEncoder {
+
+    /**
+     * Utility class.
+     */
+    private NameEncoder() {
+        throw new IllegalAccessError("Utility class");
+    }
+
+    public static String encode(String name) {
+        if ("".equals(name)) {
+            return "%00";
+        }
+        if (".".equals(name)) {
+            // just enough escaping to bypass the URL segment meaning of ".."
+            return "%2E";
+        }
+        if ("..".equals(name)) {
+            // just enough escaping to bypass the URL segment meaning of ".."
+            return "%2E.";
```

Comment:
I wonder if some web servers will treat an attempt to visit such URLs as an attempted hack and block them. I know we have had problems whereby valid URLs like `http://jenkins/descriptorById/whatever/thing/foo%2Fbar/` correctly call `Whatever.DescriptorImpl.getThing("foo/bar")` in most cases but in some configurations of Tomcat produce a 404 for supposed security reasons.